File: june14/aug31_LL4/floydcycle.cpp
#include<iostream>

using namespace std;

class Node
{
    public:
        int data;
        Node* next;

        Node()
        {
            this->data = 0;
            this->next = NULL;
        }

        Node(int data, Node* next)
        {
            this->data = data;
            this->next = next;
        }
};

void display(Node* head)
{
    for(Node* i = head; i != NULL; i = i->next)
        cout << i->data << " ";
    cout << endl;
}

void addLast(Node*& head, Node*& tail, int data)
{
    if(head == NULL)
        head = tail = new Node(data, NULL);
    else
    {
        Node* nn = new Node(data, NULL);
        tail->next = nn;
        tail = nn;
    }
}

Node* getNodeAt(Node* head, int idx)
{
    if(head == NULL)
        return NULL;
    
    int j = 0;
    for(Node* i = head; i != NULL; i = i->next)
    {
        if(j == idx)
            return i;
        j++;
    }
    return NULL;
}

//my method
void floydCycle(Node* head)
{
    Node* slow = head;
    Node* fast = head;
    do
    {
        slow = slow ->next;
        fast = fast->next->next;
    }while(fast != slow && fast != NULL && fast->next != NULL);

    if(slow != fast)
    {
        cout << "No cycle" << endl;
        return;
    }
    else
    {
        cout << "Cycle detected" << endl;
        if(slow == head)//Self loop
        {
            while(slow->next != head)
                slow = slow->next;
        }
        else
        {
            fast = head;
            while(slow->next != fast->next)
            {
                fast = fast->next;
                slow = slow->next;
            }
        }

        slow->next = NULL;//cycle removed
    }
}

void detectandRemoveLoop(Node*& head)
{
    Node* slow = head;
    Node* fast = head;
    cout<<"loop detection"<<endl;
    while(fast != NULL && fast->next != NULL)
    {
        slow = slow ->next;
        fast = fast->next->next;
        if(slow == fast)
        {
            cout << "Cycle detected" << endl;
            if(slow == head)//Self loop
            {
                while(slow->next != head)
                    slow = slow->next;
            }
            else
            {
                fast = head;
                while(slow->next != fast->next)
                {
                    fast = fast->next;
                    slow = slow->next;
                }
            }

            slow->next = NULL;//cycle removed
            display(head);
            break;
        }
    }

}

int main(int argc, char** argv)
{
    Node* head = NULL;
    Node* tail = NULL;
    addLast(head, tail, 10);
    addLast(head, tail, 20);
    addLast(head, tail, 30);
    addLast(head, tail, 40);
    addLast(head, tail, 50);
    addLast(head, tail, 60);
    addLast(head, tail, 70);
    addLast(head, tail, 80);
    addLast(head, tail, 90);
    addLast(head, tail, 100);
    tail->next = getNodeAt(head, 5);
    

    // floydCycle(head);
    // display(head);
    detectandRemoveLoop(head);
    // display(head);
}

File: feb2/mar23_LL3/floydcycle.java
package mar23;

public class floydcycle {

	class Node{
		public int data;
		public Node next;
		
		Node(int data){
			this.data=data;
		}
	}
	
	public static void display(Node node){
		for(Node temp=node;temp!=null;temp=temp.next)
			System.out.print(temp.data+" ");
		System.out.println();
	}
	
	public static void removeLoop(Node slow,Node fast){
		Node temp=fast;
		while(slow!=fast){
			slow=slow.next;
			temp=fast;
			fast=fast.next;
		}
//		System.out.println(slow.data);//intersection point
		temp.next=null;
	}
	public static boolean dar(Node node){//dar= detect and remove
		Node slow=node;
		Node fast=node;
		while(fast!=null && fast.next!=null){
			slow=slow.next;
			fast=fast.next.next;
			if(slow==fast){
				removeLoop(node,fast);
				return true;
			}
		}
		return false;
	}
	public static void main(String[] args) {
		floydcycle f=new floydcycle();
		Node n1=f.new Node(10);
		Node n2=f.new Node(20);
		Node n3=f.new Node(30);
		Node n4=f.new Node(40);
		Node n5=f.new Node(50);
		Node n6=f.new Node(60);
		Node n7=f.new Node(70);
		Node n8=f.new Node(80);
		Node n9=f.new Node(90);
		Node n10=f.new Node(100);
		
		n1.next=n2;
		n2.next=n3;
		n3.next=n4;
		n4.next=n5;
		n5.next=n6;
		n6.next=n7;
		n7.next=n8;
		n8.next=n9;
		n9.next=n10;
		n10.next=n5;
		
		System.out.println(dar(n1));
		display(n1);
		
		

	}

}
